### PR TITLE
ClusterBrainLocation and SessionBrainLocation

### PR DESCRIPTION
**Pull request by shenshan**

### CODE REVIEW

File: scripts/updates/add_level_graph_order.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from ibl_pipeline import reference
+from ibl_pipeline.ingest import reference as reference_ingest
+import datajoint as dj
+from tqdm import tqdm
+
```
Comment (eywalker) [inline scripts/updates/add_level_graph_order.py:5]: let's wrap the execution within script specific block.

File: ibl_pipeline/job/patch.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+
+table_kwargs = dict(order_by='table_order desc', as_dict=True)
+populate_kwargs = dict(
+    suppress_errors=True, display_progress=True,
+    return_exception_objects=True)
+
+
+@schema
+class Run(dj.Manual):
```
Comment (eywalker) [inline ibl_pipeline/job/patch.py:154]: Quite a cool logic but curious why you didn't turn this into a computed table?

File: scripts/updates/update_session_uuids.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                 (acquisition.Session.proj(uuid='session_uuid') &
+                  session_raw)) & behavior.TrialSet
+
+# eid non matching sessions with matching subject_uuid and session_start_time
+
+session_start_time_field = dj.U('uuid', 'session_start_time') & \
+    (alyxraw.AlyxRaw.Field & session_raw & 'fname="start_time"').proj(
+        session_start_time='cast(fvalue as datetime)')
+
+for (key, uuid, session_uuid) in tqdm(zip(*(prob_sessions *
```
Comment (eywalker) [inline scripts/updates/update_session_uuids.py:37]: if you could separate this out into two steps where you fetch first and then use tqdm, you can give more informative progress bar:

```python
vals = (prob_sessions * session_start_time_field).fetch('KEY', 'uuid', 'session_uuid')
for (key, uuid, session_uuid) in tqdm(zip(*vals), position=0, total=len(vals[0])):
```

Comment (eywalker) [reply scripts/updates/update_session_uuids.py:37]: This is to supplement the fact that `zip` operation doesn't provide for length information.

File: scripts/auto_update_trajectories.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (eywalker) [inline scripts/auto_update_trajectories.py:1]: let's wrap the operation of the script (esp part like modification of `dj.config['safemode']`) into `if __name__ == "__main__":` block.

File: Dockerfile [hunk 5]
```diff
@@ -13,3 +13,5 @@ RUN pip install scikits.bootstrap
 
 RUN pip install ibllib
 RUN pip install "git+https://github.com/ixcat/djwip.git#egg=djwip"
+
+ADD ./allen_structure_tree.csv /usr/local/lib/python3.6/dist-packages/ibllib/atlas
```
[the remark this answers is not in the document]
Comment (shenshan) [reply Dockerfile:17]: This is required by the ibllib, they did not put the file in place.

File: ibl_pipeline/behavior.py [hunk 6]
```diff
@@ -498,187 +497,113 @@ def make(self, key):
...
+        ses_path = alf.io.get_session_path(files[0])
+        trials = alf.io.load_object(
+            ses_path.joinpath('alf'), '_ibl_trials')
+
+        status = (CompleteTrialSession & key).fetch1()
 
         lab_name = (subject.SubjectLab & key).fetch1('lab_name')
-        if stim_on_times_status != 'Missing':
+        if status['stim_on_times_status'] != 'Missing':
             if lab_name == 'wittenlab':
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/behavior.py:528]: This is for some historical reason. I could probably remove them now.

File: ibl_pipeline/ingest/ingest_utils.py [hunk 7]
```diff
@@ -4,23 +4,23 @@
...
     if fresh:
-        target_table.insert(src_table)
+        target_table.insert(src_table, **kwargs)
     else:
         try:
             target_table.insert(src_table - target_table.proj(),
-                                skip_duplicates=True)
-        except:
+                                skip_duplicates=True, **kwargs)
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/ingest/ingest_utils.py:17]: No, not really, but PEP8 already complains if I just do Exception.

File: ibl_pipeline/behavior.py [hunk 8]
```diff
@@ -4,11 +4,14 @@
...
+import warnings
 from . import reference, subject, acquisition, data
 try:
     from oneibl.one import ONE
-except:
-    # TODO: consider issuing a warning about not being able to perform ingestion without ONE access
+    import alf.io
+    one = ONE()
+except Exception:
+    warnings.warn('ONE not installed, cannot use populate')
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/behavior.py:14]: Good point. Fixed.

File: ibl_pipeline/behavior.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/behavior.py:528]: Ok removed.

File: ibl_pipeline/ephys.py [hunk 9]
```diff
@@ -340,6 +284,13 @@ class Metrics(dj.Part):
         metrics:                    longblob    # a dictionary with fields of metrics, depend on the clustering method
         """
 
+    class Ks2Label(dj.Part):
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/ephys.py:287]: Good suggestion. Added comments to all ephys tables.

File: ibl_pipeline/ephys.py [hunk 10]
```diff
@@ -340,6 +284,13 @@ class Metrics(dj.Part):
...
+        definition = """
+        -> master
+        ---
+        ks2_label       : enum('good', 'mua')
+        """
+
 
 @schema_groupshare
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/ephys.py]: Yes you're right. The table is not used, so I removed that table.

File: ibl_pipeline/ephys.py [hunk 11]
```diff
@@ -461,7 +381,7 @@ def make(self, key):
             spike_times)
 
         trial_spks = []
-        for itrial, trial_key in tqdm(enumerate(trial_keys)):
+        for itrial, trial_key in tqdm(enumerate(trial_keys), position=0):
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/ephys.py:383]: `position=0` is to keep the progress bar in the same position while looping through (this is my understanding), to keep the log readable. It helps on some cases but not always.

File: ibl_pipeline/group_shared/wheel.py [hunk 12]
```diff
@@ -0,0 +1,278 @@
+import datajoint as dj
+from ibl_pipeline import acquisition, behavior
+from oneibl.one import ONE
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/group_shared/wheel.py:3]: You're right. This file was copied from Miles, just for the sake of patching. I might eventually keep it as a dependent package. I probably won't change this file for the moment.

File: ibl_pipeline/histology.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    def make(self, key):
+        channel_raw_inds, channel_local_coordinates = \
+            (ephys.ChannelGroup & key).fetch1(
+                'channel_raw_inds', 'channel_local_coordinates')
+        channel = (ephys.DefaultCluster & key).fetch1('cluster_channel')
+        if channel in channel_raw_inds:
+            channel_coords = np.squeeze(
+                channel_local_coordinates[channel_raw_inds == channel])
+        else:
+            return
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/histology.py:104]: Yes. If clusters.channels or channels.localCoordinates got patched, rerunning populate will take care the clusters. But also likely the clusters of the whole sessions will be reprocessed, depending on why the those data were not correct.

File: notebooks/notebooks_tutorial/202007_neuro_match_academy/paper_behavior_functions.py [hunk 14]
```diff
@@ -0,0 +1,237 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (shenshan) [reply notebooks/notebooks_tutorial/202007_neuro_match_academy/paper_behavior_functions.py:1]: This is user functions. Could definitely skip :)

File: ibl_pipeline/job/patch.py (hunk 2 above)
Comment (shenshan) [reply ibl_pipeline/job/patch.py:154]: The reasons are:
1. I called a lot of populate in the make. It wouldn't work as a computed table.
2. I would like to track the status in real time, such as which table is populating, which ones are finished. I would want the transactions to happen in real time, not till the end of the program.

File: scripts/updates/update_session_uuids.py (hunk 3 above)
Comment (shenshan) [reply scripts/updates/update_session_uuids.py]: Wow, cool.

File: scripts/dbtest/table_names.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
+'''
+Create global variables for table names
+'''
+
+
+def init():
+
+    global REF_TABLES, SUBJECT_TABLES, ACTION_TABLES, \
```
[the remark this answers is not in the document]
Comment (shenshan) [reply scripts/dbtest/table_names.py:8]: ok fixed as suggested.

File: scripts/dbtest/ingest_until_session.sh [hunk 16]
```diff
@@ -0,0 +1,8 @@
+python ingest_alyx_raw_dbtest.py
```
[the remark this answers is not in the document]
Comment (shenshan) [reply scripts/dbtest/ingest_until_session.sh:1]: Good idea.

File: scripts/updates/delete_bad_aligned_spikes.py [hunk 17]
```diff
@@ -0,0 +1,11 @@
+import datajoint as dj
+from ibl_pipeline import ephys
+from tqdm import tqdm
+
+dj.config['safemode'] = False
+
+clusters = ephys.DefaultCluster & \
+    (ephys.AlignedTrialSpikes & 'event="movement"')
+
+for key in tqdm(clusters.fetch('KEY'), position=0):
```
[the remark this answers is not in the document]
Comment (shenshan) [reply scripts/updates/delete_bad_aligned_spikes.py]: Good suggestions, I add the line to most of the scripts. will do the rest whenever have a chance using them.

File: ibl_pipeline/ingest/ingest_alyx_raw.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+import numpy as np
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_alyx_entries(filename=None, models=None,
+                     exclude=[]):
+
+    exclude_list = ['auth.group', 'sessions.session',
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/ingest/ingest_alyx_raw.py]: Good to know. Changed.

File: ibl_pipeline/ingest/ingest_alyx_raw.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+                    'authtoken.token',
+                    'experiments.brainregion'] + exclude
+
+    if not filename:
+        dir_name = path.dirname(__file__)
+        if len(sys.argv) < 2:  # no arguments given
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/ingest/ingest_alyx_raw.py:33]: Good point. I moved this part below `if __name__=="__main__"` and pass in the filename directly to this function.

File: ibl_pipeline/ingest/ingest_alyx_raw.py [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+                else:
+                    key_field['value_idx'] = 0
+                    key_field['fvalue'] = str(field_value)
+                    ib_part.insert1(key_field)
+
+                if ib_part.flush(skip_duplicates=True, chunksz=10000):
+                    logger.debug('Inserted 10000 raw field tuples')
+                    # print('Inserted 10000 raw field tuples')
+        except Exception as e:
+            print('Problematic entry:{}'.format(ikey))
```
[the remark this answers is not in the document]
Comment (shenshan) [reply ibl_pipeline/ingest/ingest_alyx_raw.py:128]: Not sure about this. Just would like to keep the log on this short.

File: scripts/dbtest/ingest_until_session.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (shenshan) [reply scripts/dbtest/ingest_until_session.sh:1]: Great suggestion. It's a bit too much to add to every script right now, but I got the idea. Will add this line whenever actively working on the scripts.

File: ibl_pipeline/job/patch.py (hunk 2 above)
Comment (eywalker) [reply ibl_pipeline/job/patch.py:154]: Makes a lot of sense. Thanks for the clarification!

**Review by eywalker — APPROVED** [review]

**Pull request merged**